src/protocol/dot_product.rs

```diff
@@ -0,0 +1,173 @@
+        record_id: RecordId,
+        mut input_stream: Pin<Box<dyn Stream<Item = PairOfShares<F>>>>,
+    ) -> Result<Replicated<F>, BoxError> {
+        let prss = &ctx.prss();
+        let (s_left, s_right): (F, F) = prss.generate_fields(record_id);
+
+        let mut u_right = s_right - s_left;
+
+        while let Some(item) = input_stream.next().await {
+            let (a_share, b_share) = item;
+
+            let (a_left, a_right) = a_share.as_tuple();
+            let (b_left, b_right) = b_share.as_tuple();
+            u_right += (a_left + a_right) * (b_left + b_right) - a_right * b_right;
+        }
+
+        // send our `u_i+1` value to the helper on the right
+        let channel = ctx.mesh();
+        channel
+            .send(
+                channel.identity().peer(Direction::Right),
+                record_id,
+                UValue { u: u_right },
+            )
+            .await?;
+
+        // receive `u_i` value from helper to the left
+        let UValue { u: u_left } = channel
+            .receive(channel.identity().peer(Direction::Left), record_id)
+            .await?;
+
+        Ok(Replicated::new(u_left, u_right))
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use crate::error::BoxError;
+        use crate::field::Field;
+        use crate::field::Fp31;
+        use crate::protocol::dot_product::stream::accumulate_dot_product;
+        use crate::protocol::{QueryId, RecordId};
+        use crate::test_fixture::{
+            logging, make_contexts, make_world, share, validate_and_reconstruct,
+        };
+        use futures_util::stream;
+        use proptest::prelude::Rng;
+
+        #[tokio::test]
+        async fn supports_stream_of_secret_shares() -> Result<(), BoxError> {
+            logging::setup();
+
+            let mut rng = rand::thread_rng();
+
+            let mut vec1 = Vec::with_capacity(100);
+            let mut vec2 = Vec::with_capacity(100);
+            let mut vec3 = Vec::with_capacity(100);
+
+            let mut correct_answer = Fp31::ZERO;
+
+            for _ in 0..100 {
+                let a_i = Fp31::from(rng.gen::<u128>());
+                let b_i = Fp31::from(rng.gen::<u128>());
+                let a_i_shares = share(a_i, &mut rng);
+                let b_i_shares = share(b_i, &mut rng);
+                vec1.push((a_i_shares[0], b_i_shares[0]));
+                vec2.push((a_i_shares[1], b_i_shares[1]));
+                vec3.push((a_i_shares[2], b_i_shares[2]));
+                correct_answer += a_i * b_i;
+            }
+
+            // setup helpers
+            let world = make_world(QueryId);
+            let context = make_contexts(&world);
+            let record_id = RecordId::from(0);
+            let step = "some_step".to_string();
+
+            let result_shares = tokio::try_join!(
+                accumulate_dot_product(
+                    context[0].narrow(&step),
+                    record_id,
+                    Box::pin(stream::iter(vec1)),
```

Comment:
if you move away from `Pin`, you can drop the `Box::pin` here and pass the `Stream` directly

Reply:
That's awesome. Thank you so much @thurstonsand !